src/widgets/helper/NotebookTab.cpp

```diff
@@ -381,9 +517,82 @@ void NotebookTab::setHighlightState(HighlightState newHighlightStyle)
     this->update();
 }
 
-HighlightState NotebookTab::highlightState() const
+void NotebookTab::updateHighlightState(HighlightState newHighlightStyle,
+                                       const ChannelView &channelViewSource,
+                                       const MessagePtr &message)
 {
-    return this->highlightState_;
+    if (this->isSelected())
+    {
+        assert(this->highlightSources_.empty());
+        assert(this->highlightState_ == HighlightState::None);
+        return;
+    }
+
+    if (!this->shouldMessageHighlight(channelViewSource, message))
+    {
+        return;
+    }
+
+    if (!this->highlightEnabled_ &&
+        newHighlightStyle == HighlightState::NewMessage)
+    {
+        return;
+    }
+
+    // message is highlighting unvisible tab
+
+    auto channelViewId = channelViewSource.getID();
+
+    switch (newHighlightStyle)
+    {
+        case HighlightState::Highlighted:
+            // override lower states
+            this->highlightSources_.insert_or_assign(channelViewId,
+                                                     newHighlightStyle);
+        case HighlightState::NewMessage: {
+            // only insert if no state already there to avoid overriding
+            if (!this->highlightSources_.contains(channelViewId))
+            {
+                this->highlightSources_.emplace(channelViewId,
+                                                newHighlightStyle);
+            }
+            break;
+        }
+        case HighlightState::None:
+            break;
+    }
+
+    if (this->highlightState_ == newHighlightStyle ||
+        this->highlightState_ == HighlightState::Highlighted)
+    {
+        return;
+    }
+
+    this->highlightState_ = newHighlightStyle;
+    this->update();
+}
+
+bool NotebookTab::shouldMessageHighlight(const ChannelView &channelViewSource,
+                                         const MessagePtr &message) const
```

Comment:
This function is essentially: `isMessageInAnyCurrentlyRenderedChannelView` - could it & its logic be moved to ChannelView instead (maybe as an anonymous function) so `tabHighlightRequested` doesn't need the `message` param, and a tab highlight always means a highlight will take place (except if the tab is selected)?

Reply:
I've looked through it and I believe in here:

https://github.com/Chatterino/chatterino2/blob/05ab3b27e783c33aceaed78f283a6b05aeed54e5/src/widgets/helper/NotebookTab.cpp#L585-L588

the channelViewSource.shouldIncludeMessage is superfluous, the way the message flows indicates it already was checked.

But the problem is that we need to figure out if _other_ visible splits can see this message.

Unfortunately I don't see a way to remove the message param, since the handling happens both asynchronously and most notably - independently.

Reply:
The only need for this message would be if we want the underlying channel to not highlight if a message appears and is also shown in visible filtered split.


What should then be done?
1) message appearing in currently visible filtered channel should not highlight other splits that are of same underlying channel but _unfiltered_ -> this needs the message there
2) they should highlight -> this does not need to check any message

Reply:
Currently with the switch to identity based checking and treating each filtered channel as its own different channel, the message does not need to be there. So I removed it now.

There is a possibility for message checking if we want to make messages that appear in currently visible filtered split to not highlight some other splits where the message also appears. But it only makes sense for a message not to appear in unfiltered split, as other filtered splits are set by the user to show the message for some reason.

Reply:
Awesome :+1: This looks better, even if NotebookTab has some logic I don't want it to have, this is acceptable